Fix bug #23712: Mage of Light halo not drawing correctly

jyrkive opened the pull request:
The bug was caused by overly aggressive performance optimizations.

When a unit is moving, it's replaced with a temporary unit. If the moving unit is a Mage of Light, its halo is recreated. The recreated halo is originally placed in the top-left corner of the screen (exactly the position where it actually stayed in practice).

The game attempts to correct the position of the halo on the next frame by calling `halo::manager::set_location()`, that eventually calls `halo_impl::effect::set_location()`. Since the overlayed hexes (`overlayed_hexes_`) aren't known after moving the halo, `set_location()` clears the list of overlayed hexes. The idea is that the overlayed hexes will be determined when the halo is drawn next time, in `halo_impl::effect::render()`.

`Halo_impl::render()` only draws invalidated haloes.

Haloes are invalidated in `halo_impl::unrender()`. It only invalidates haloes which overlay invalidated hexes.

But the list of overlayed hexes is empty because the halo hasn't been drawn in the correct position yet. Thus, the halo won't be invalidated, or later drawn. The result is a kind of deadlock: the halo is never drawn in the correct position because it's never invalidated, and it's never invalidated because it's never drawn in the correct position.

My new code detects the situation in `halo_impl::unrender()` and invalidates haloes whose positions are not known.

Wedge009 commented:
Thanks for your efforts! I can confirm that the halo does again appear around the mage upon moving, and the explanation and code changes seem sound to me. However, unfortunately, the incorrect haloes at the top-left corner are still appearing and, it appears, are being stacked upon each other with every move the mage makes. I don't have time to look into this more thoroughly myself, but do you think you can see if there's anything you can find with regards to this last problem? Thanks again!

jyrkive commented:
Oh, right. I didn't notice that the haloes continue to appear in the top-left corner as well. I'll investigate. (My guess is that the halo isn't "unrendered" properly.)

jyrkive commented:
I committed a fix that calls `halo_impl::effect::unrender()` in `halo_impl::effect::set_location()`. It keeps the haloes from showing in the top-left corner. However, I'm not sure if `unrender()` is a function that can be safely called at any time (like `set_location()` should probably be). Can anyone who knows the rendering pipeline better than me comment on that?

aginor commented:
Hi @jyrkive, thanks for your contribution :)

Unrender can theoretically be called from any location, but it is rather finicky as it restores a copied version of the surface. I'd prefer a different solution if possible as all calls to unrender-like functionality should be removed so we can transition to openGL and hardware acceleration.
I'd however be happy to accept the patch as-is.

Could you please add yourself to the authors list too and squash that commit into the previous?

Wedge009 commented:
There's still a bit of flickering as the halo is rendered and unrendered, but it does the job. Thanks for your help.

jyrkive commented:
I mentioned in IRC that probably the easiest way to eliminate the flicker would be by making the unit not be replaced with a temporary unit when it steps into a new hex.

The code that recreates the temporary unit is here:

```c++
void unit_mover::proceed_to(unit_ptr u, size_t path_index, bool update, bool wait)
{
	// Nothing to do here if animations cannot be shown.
	if ( !can_draw_ || !animate_ )
		return;

	// Handle pending visibility issues before introducing new ones.
	wait_for_anims();

	if ( update  ||  !temp_unit_ptr_ )
		// Replace the temp unit (which also hides u and shows our temporary).
		replace_temporary(u);
	else
	{
		// Just switch the display from the real unit to our fake one.
		temp_unit_ptr_->set_hidden(false);
		u->set_hidden(true);
	}
```

In other words, the unit will be replaced only if the `update` parameter is set to true.

Calling code is in `unit_mover::do_move()`:

```c++
// Show this move.
const size_t current_tracking = resources::game_events->pump().wml_tracking();
animator.proceed_to(move_it_.get_shared_ptr(), step_to - begin_,
                                  current_tracking != do_move_track_, false);
do_move_track_ = current_tracking;
disp.redraw_minimap();
```

The documentation for `game_events::t_pump::wml_tracking()` states that the function is used to detect if WML or Lua scripts have been executed. Those scripts can do who-knows-what to the moving unit, and so it makes sense for the unit mover to recreate the temporary unit between steps. Thus, it looks like not recreating the unit on every step is not an option.

Next, I'll explore the possibility to share the halo between the temporaries and the real unit (i.e. let the unit mover recreate the unit, but not the halo).

jyrkive commented:
Some progress. Sharing the halo was very easy to implement after all (a one-line change), but sadly it didn't fix the flicker.

I added some logging. [The log](https://gist.github.com/jyrkive/2cb4ae93d84b32b677a985f8a0398da9) shows that at some points during the animation, the halo is unrendered on frame N and only subsequently rendered on frame N+1.

> 20160626 18:02:15 debug halo: Unrender halo on frame 115
> 20160626 18:02:15 debug halo: Render halo on frame 116

> 20160626 18:02:15 debug halo: Unrender halo on frame 125
> 20160626 18:02:15 debug halo: Render halo on frame 126

> 20160626 18:02:15 debug halo: Unrender halo on frame 135
> 20160626 18:02:15 debug halo: Render halo on frame 136

> 20160626 18:02:15 debug halo: Unrender halo on frame 145
> 20160626 18:02:15 debug halo: Render halo on frame 146

> 20160626 18:02:15 debug halo: Unrender halo on frame 155
> 20160626 18:02:15 debug halo: Render halo on frame 156

Once every ten frames.

I need to add heavier logging to figure out what's going on...

jyrkive commented:
It turned out that sharing the halo is too hard after all. The initial temporary unit shared the halo with the real unit, but right after that the real unit lost its reference to the halo (because the unit was hidden). Subsequent temporary units are created off the real unit, not the previous temporary, and therefore the new subsequent unit didn't get a reference to the shared halo. It ended up creating its own halo in the end.

Getting sharing the halo to work would have required an ability for invisible halos to exist. To track if halos are visible or not, every halo would need references to all the units which use it. I realized that approach was way too ugly.

So, I took the other option: make sure that when the halo is recreated, it will be drawn already in the next frame, and in the correct position. It turned out to be surprisingly easy. The above log is misleading because the haloes were actually unrendered **between** frames (for example, frame 115 had already been blit to screen when the halo was unrendered).

The only reason why the animation was flickering was because the halo was initially placed in the top-left corner. Placing it in the right place from the start was all it took to fix the flicker. :+1: 

If you're wondering about the change in the constructor of `halo_impl::effect`, that change doesn't change anything regarding functionality. The constructor calls `set_location()` which assigns the final values of `x_` and `y_`. I changed the initializer list because having `x_` and `y_` being assigned the passed values made me think that the constructor wouldn't change the values afterwards (and, therefore, that the values wouldn't be adjusted like `set_location()` does).

@aginor My name is already in the authors list. I squashed my commits together.

Wedge009 commented:
I was wondering if it should be simple to just correct the co-ordinates of the halo to begin with, thanks! I've given this a test, plus checked the Elvish Shyde's halo as well, and there seems to be no obvious regressions while also fixing the flickering and improper location.

I give my recommendation for this to be merged.

aginor commented:
Yes, it looks great. @jyrkive, thanks for doing a great job!
@Wedge009, thanks for helping out with the testing.

newfrenchy83 commented:
placehalo_man. set_location (ac.unit_halo_, x, y - height_adjust) after ac.unit_halo_ = halo_man. add ( 0, 0, u. image_halo ()+u. TC_image_mods (), map_loca... was simply but functionnal without artefact in drawing halo. I had process the change in my copy of source code in 2015 but my social phobie et my internet difficulty are blocked me.

jyrkive commented:
@newfrenchy83 The only line in this pull request that was necessary to fix the bug was calling `halo_man.add()` with `x` and `y - height_adjust`. The other changes are optional, they fix smaller bugs.

It is true that adding a call for `halo_man.set_location()` right after `halo_man.add()` would fix the issue as well. However, there isn't any reason why we would revert the earlier fixes and replace them with yours.

newfrenchy83 commented:
i trying to explain jyrlive.When you place two mage of light adjacent, look the artifact where the halos crossing, after try your soluce, i revert
halo.cpp and drawer.cpp and replace by my soluve wjat i using before you commit, the artifact no appear in my version.
the other files resolve another bug, the halo peristant when units die or modified( and i not commits change in unstore_unit find vacant=no using erase_unit). i hope you will undertand mu explain. i like what someone test the two soluce and compare the crossing halo after move two mage of light adjacent, i not commits my alternate soluce wirhout reason, please. i not wantef insult your work

jyrkive commented:
Okay, that's an explanation. I guess I'll need to test the behavior myself.

Based on what you told, the crossing halo issue is likely related to one or two of the changes I made in halo.cpp. I'd still like to understand why exactly, and it would probably be best if I investigated the behavior myself.
